pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,185 @@
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	containerData, err := c.ContainerdTaskToContainerData(response.Process)
+	if err != nil {
+		return nil, err
+	}
+	return containerData, nil
+}
+
+func (c *ContainerdClient) GetContainerDetails(containerID string) (*runtimeclient.ContainerDetailsData, error) {
+	response, err := c.client.TaskService().Get(context.TODO(), &tasks.GetRequest{
+		ContainerID: containerID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	proc := response.Process
+
+	containerData, err := c.ContainerdTaskToContainerData(proc)
+	if err != nil {
+		return nil, err
+	}
+
+	containers, err := c.client.Containers(context.TODO(), fmt.Sprintf("id==%s", containerID))
+	if err != nil {
+		return nil, err
+	}
+	if len(containers) != 1 {
+		return nil, fmt.Errorf("expected 1 container with id %q, got %d", containerID, len(containers))
+	}
+
+	spec, err := containers[0].Spec(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	mountData := make([]runtimeclient.ContainerMountData, len(spec.Mounts))
+	for i := range spec.Mounts {
+		mount := spec.Mounts[i]
+		mountData[i] = runtimeclient.ContainerMountData{
+			Source:      mount.Source,
+			Destination: mount.Destination,
+		}
+	}
+
+	return &runtimeclient.ContainerDetailsData{
+		ContainerData: *containerData,
+		Pid:           int(proc.Pid),
+		CgroupsPath:   spec.Linux.CgroupsPath,
+		Mounts:        mountData,
 	}, nil
 }
+
+// Convert the state from container status to state of runtime client.
+func processStatusStateToRuntimeClientState(status task.Status) (runtimeClientState string) {
+	switch status {
+	case task.StatusCreated:
+		runtimeClientState = runtimeclient.StateCreated
+	case task.StatusRunning:
+		runtimeClientState = runtimeclient.StateRunning
+	case task.StatusStopped:
+		runtimeClientState = runtimeclient.StateExited
+	default:
+		runtimeClientState = runtimeclient.StateUnknown
+	}
+	return
+}
+
+func guessContainerName(container containerd.Container) string {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return container.ID()
+	}
+
+	if nerdctlName, ok := labels[ContainerNameNerdctl]; ok {
+		return nerdctlName
+	} else if k8sName, ok := labels[ContainerNameK8s]; ok {
+		return k8sName
+	}
+
+	return container.ID()
+}
+
+func (c *ContainerdClient) ContainerdTaskToContainerData(proc *task.Process) (*runtimeclient.ContainerData, error) {
+	containers, err := c.client.Containers(context.TODO(), fmt.Sprintf("id==%s", proc.ID))
+	if err != nil {
+		return nil, err
+	}
+	if len(containers) != 1 {
+		return nil, fmt.Errorf("expected 1 container with id %q, got %d", proc.ContainerID, len(containers))
+	}
+
+	labels, err := containers[0].Labels(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	containerData := &runtimeclient.ContainerData{
+		ID:      proc.ID,
+		Name:    guessContainerName(containers[0]),
+		State:   processStatusStateToRuntimeClientState(proc.Status),
```

Comment:
Getting the task and then getting the container means we need two, instead of one request, to enrich each container. It seems to be needed to get the `State`, but I am not sure it is worth it. `nerdctl` is directly using `client.Containers()` for the [ps command](https://github.com/containerd/nerdctl/blob/667307635d06f38616699a0d52f96f75389fcefa/cmd/nerdctl/ps.go#L91) and it somehow infers the status. Did you check it?

Reply:
Did not check it before.
As far as i can see from looking at the code they also read the `Task` in [their ContainerStatus formatter function](https://github.com/containerd/nerdctl/blob/667307635d06f38616699a0d52f96f75389fcefa/pkg/formatter/formatter.go#L46)

Reply:
I see. However, for our use case, I would still try to avoid that additional call and have to deal with `Task`. 

I checked what we do with the `State`, and we only use it for discarding non-running containers after calling `GetContainers`. Instead, for `GetContainer` and `GetContainerDetails`, we don't use it at all. Maybe from `GetContainers`, we could filter the `client.Containers` call to return only the running containers? Is it possible?

Reply:
The filter string bases on the protobuf fields. They do not appear to give us any informations regarding the state of the container or its task :(
https://github.com/containerd/containerd/blob/main/api/services/containers/v1/containers.pb.go

Reply:
Did you check if `client.Containers` maybe already returns only the running containers?

Reply:
Unfortunately, it returns all containers

Reply:
> To avoid such redundancy and improve the performance, we could move the logic of steps 1 and 2 inside the runtime implementations.

Yes, I think we need that. It seems to me the needed for Container and ContainerDetails comes from docker and CRI API. 

> What do you think?

I think it's fine to have 4 calls for the time being, but we should keep an eye to improve it later on.